R/graded.R

```diff
@@ -451,39 +552,94 @@ fail_if <- function(
   }
 }
 
-#' @describeIn graded Signal a _failing_ grade if mistakes are detected in the
-#'   submitted code.
-#'   
+#' Signal a failing grade if mistakes are detected in the submitted code
+#' 
+#' @description
+#' `fail_if_code_feedback()` uses [code_feedback()] to detect if there are
+#' differences between the user's submitted code and the solution code (if
+#' available). If the exercise does not have an associated solution, or if there
+#' are no detected differences between the user's and the solution code, no
+#' grade is returned.
+#' 
+#' See [graded()] for more information on \pkg{gradethis} grade-signaling
+#' functions.
+#' 
+#' @examples
+#' # Suppose the exercise prompt is to generate 5 random numbers, sampled from
+#' # a uniform distribution between 0 and 1. In this exercise, you know that
+#' # you shouldn't have values outside of the range of 0 or 1, but you'll
+#' # otherwise need to check the submitted code to know that the student has
+#' # chosen the correct sampling function.
+#' 
+#' grader <-
+#' # ```{r example-check}
+#'   grade_this({
+#'     fail_if(length(.result) != 5, "I expected 5 numbers.")
+#'     fail_if(
+#'       any(.result < 0 | .result > 1), 
+#'       "I expected all numbers to be between 0 and 1."
+#'     )
+#'     
+#'     # Specific checks passed, but now we want to check the code.
+#'     fail_if_code_feedback()
+#'     
+#'     # All good!
+#'     pass()
+#'   })
+#' # ```
+#' 
+#' .solution_code <- "
+#' # ```{r example-check}
+#'   runif(5)
+#' # ```
+#' "
+#' 
+#' # Not 5 numbers...
+#' grader(mock_this_exercise(runif(1), !!.solution_code))
+#' 
+#' # Not within [0, 1]...
+#' grader(mock_this_exercise(rnorm(5), !!.solution_code))
+#' 
+#' # Passes specific checks, but hard to tell so check the code...
+#' grader(mock_this_exercise(runif(5, 0.25, 0.75), !!.solution_code))
+#' grader(mock_this_exercise(rbinom(5, 1, 0.5), !!.solution_code))
+#' 
+#' # Perfect!
+#' grader(mock_this_exercise(runif(n = 5), !!.solution_code))
+#' 
 #' @inheritParams code_feedback
-#'
+#' @inheritParams graded
+#' @inheritDotParams graded
+#' 
+#' @return Signals an incorrect grade with feedback if there are differences
+#'   between the submitted user code and the solution code. If solution code is
+#'   not available, no grade is returned.
+#'
+#' @template graded-family
 #' @export
 fail_if_code_feedback <- function(
-  user_code = missing_arg(),
-  solution_code = missing_arg(),
   message = NULL,
+  user_code = .user_code,
+  solution_code = .solution_code,
   ...,
   env = parent.frame(),
   hint = TRUE,
   encourage = getOption("gradethis.fail.encourage", FALSE),
   allow_partial_matching = getOption("gradethis.allow_partial_matching", TRUE)
 ) {
-  if (is_missing(user_code)) {
+  if (identical(rlang::enexpr(user_code), rlang::expr(.user_code))) {
```

Comment:
I understand the motivation here, but I do not like the change.

I don't have a solution that works with both "being able to pass through a missing argument" and "displaying a _comfortable_ default value". 

Let's accept this change for now, but know it comes at the cost of being extendable.

Reply:
Same with `solution_code` below

Reply:
whoops yeah this should use the new placeholder system, I missed this during that refactor. thanks!